openassessment/data.py

```diff
@@ -959,36 +952,204 @@ def _download_file_by_key(cls, key):
+        all information needed to build the submission file path.
+        """
+
+        blocks = _get_course_blocks(course_id)
+
+        path_info = {}
+
+        def children(usage_key):
+            for index, child in enumerate(blocks.get_xblock_field(usage_key, 'children') or [], 1):
+                yield index, blocks.get_xblock_field(child, 'display_name'), child
+
+        for section_index, section_name, section in children(blocks.root_block_usage_key):
+            for sub_section_index, sub_section_name, sub_section in children(section):
+                for unit_index, unit_name, unit in children(sub_section):
+                    for block_index, block_name, block in children(unit):
+                        if block.block_type == "openassessment":
+                            path_info[str(block)] = {
+                                "section_index": section_index,
+                                "section_name": section_name,
+                                "sub_section_index": sub_section_index,
+                                "sub_section_name": sub_section_name,
+                                "unit_index": unit_index,
+                                "unit_name": unit_name,
+                                "ora_index": block_index,
+                                "ora_name": block_name,
+                            }
+
+        return path_info
+
+    @classmethod
+    def _map_student_ids_to_path_ids(cls, all_submission_information):
+        """
+        Builds a mapping between anonymized student ids and their identifier for
+        submission filename.
+
+        The identifier can take three different forms:
+        - External ID of `mb_coaching` type.
+        - edX username, if external ID is absent.
+        - Anonymized username, if `ENABLE_ORA_USERNAMES_ON_DATA_EXPORT` feature is disabled.
+        """
+        from openedx.core.djangoapps.external_user_ids.models import ExternalId
+
+        student_ids = [item[0]["student_id"] for item in all_submission_information]
+
+        User = get_user_model()
+
+        users = _use_read_replica(
+            User.objects.filter()
+            .annotate(
+                student_id=F("anonymoususerid__anonymous_user_id"),
+                path_id=Coalesce(
+                    Subquery(
+                        ExternalId.objects.filter(
+                            user=OuterRef("pk"), external_id_type__name="mb_coaching"
+                        ).values("external_user_id")
+                    ),
+                    F(
+                        "username"
+                        if _usernames_enabled()
+                        else "anonymoususerid__anonymous_user_id"
+                    ),
+                    output_field=CharField(),
+                ),
+            )
+            .values("student_id", "path_id")
+        )
+
+        return {user["student_id"]: user["path_id"] for user in users}
```

Comment:
Note: this wasn't covered with tests as well, since it relies on `ExternalId` model from Open edX core.